include protobuf service/method comments as java/scaladoc

bjaglin opened the pull request:
Fixes https://github.com/akka/akka-grpc/issues/432

bjaglin commented on codegen/src/main/scala/akka/grpc/gen/javadsl/Method.scala:
```diff
@@ -50,13 +55,24 @@ final case class Method(
 object Method {
   def apply(descriptor: MethodDescriptor): Method = {
+
+    // replicates logic from ScalaPB's DescriptorImplicits.MethodDescriptorPimp
```
it's quite horrible to have to inline that here and below but since the Java bindings don't have access to ScalaPB's pimp classes, I am not sure how it could be made better?

bjaglin replied:
I can try your suggestion indeed. I can't tell what's your final take on it though, should we push for that? Or you want other opinions?

> so some of the logic in that utility might not apply for the Java code
Is there enough java test coverage to rely on CI for that?

raboof replied:
I don't have a strong preference, so it's up to you as contributor which way to go.

Our test coverage is sufficient that I have confidence that this feature would not break anything, but a comment explaining the situation might be useful to prevent confusion when this code is touched later on.

bjaglin replied:
I have added a separate commit to use ScalaPB's `DescriptorImplicits` in a very limited and documented scope of the Java stub generation code

bjaglin commented on plugin-tester-java/src/main/protobuf/helloworld.proto:
```diff
@@ -11,8 +11,13 @@ service GreeterService {
     rpc SayHello (HelloRequest) returns (HelloReply) {}
 
+    // Comment spanning
+    // on several lines
```
I initially added these thinking I would could test non-simple comments, but I realized there was nothing at the moment in test suite that could allow to inspect metadata of the generated source files. I left it here as a passive documentation.

bjaglin commented on project/plugins.sbt:
```diff
@@ -15,7 +15,7 @@ addSbtPlugin("org.foundweekends" % "sbt-bintray" % "0.5.4")
 libraryDependencies += "org.scala-sbt" %% "scripted-plugin" % sbtVersion.value
 
-libraryDependencies += "com.thesamet.scalapb" %% "compilerplugin" % "0.8.4"
+libraryDependencies += "com.thesamet.scalapb" %% "compilerplugin" % "0.9.0-RC1"
```
first release cut out with https://github.com/scalapb/ScalaPB/pull/534

raboof replied:
It seems M2 is now out (and newer), perhaps switch to that. We might want to wait with merging before 0.9.0 is released, though.

bjaglin replied:
I just amended the commit with M2. Up to you to decide if we hold the merge until a final release is cut.

raboof replied:
I just noticed in another PR (#518) that 0.9.0-M2 does not work with jdk8 - I suspect this is unintended, so let's wait for 0.9.0

johanandren commented:
OK TO TEST

bjaglin commented:
Any interest/feedback for this?

raboof commented:
Hi! Sorry for not responding sooner - I think this would be a great improvement, but haven't found the time to look into the PR in detail yet - definitely looking forward to it!

raboof commented on codegen/src/main/scala/akka/grpc/gen/javadsl/Method.scala:
```diff
@@ -50,13 +54,23 @@ final case class Method(
   def apply(descriptor: MethodDescriptor): Method = {
+
+    val comment = {
+      // Use ScalaPB's pimp classes to avoid replicating the logic for comment extraction
```
(I personally prefer just 'implicit class' over 'pimp class', but no big deal)

raboof commented on codegen/src/main/scala/akka/grpc/gen/javadsl/Method.scala:
```diff
@@ -50,13 +54,23 @@ final case class Method(
+    val comment = {
+      // Use ScalaPB's pimp classes to avoid replicating the logic for comment extraction
+      // Note that this be problematic if/when ScalaPB uses scala-specific stuff to do that
+      implicit val ops = new DescriptorImplicits(GeneratorParams(), descriptor.getFile.getDependencies.asScala :+ descriptor.getFile)
```
Ah, you create a new empty set of parameters here instead of passing them through from the plugin. For determining the comment that's sufficient, so since this is so self-contained and documented this seems fine!

raboof commented:
scalapb 0.9.0 is now out, so this PR is unblocked and can be rebased & merged!

bjaglin commented:
I just rebased against master which has scalapb 0.9.0 and squashed in one commit with the following changes since last review:
- `pimp` wording -> `implicit`
- use Javadoc indentation for Java and Scala (the "Scaladoc" template for Scala had some heading whitespace inconsistencies)

GreeterService.scala
```
/**
 *#services
 * The greeting service definition.
 */
trait GreeterService {
  
  /**
   * Sends a greeting
   */
  def sayHello(in: example.myapp.helloworld.grpc.HelloRequest): scala.concurrent.Future[example.myapp.helloworld.grpc.HelloReply]
  
  /**
   * Comment spanning
   * on several lines
   */
  def itKeepsTalking(in: akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed]): scala.concurrent.Future[example.myapp.helloworld.grpc.HelloReply]
  
  /**
   *
   * C style comments
   */
  def itKeepsReplying(in: example.myapp.helloworld.grpc.HelloRequest): akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloReply, akka.NotUsed]
```

GreeterService.java
```
// Generated by Akka gRPC. DO NOT EDIT.
package example.myapp.helloworld.grpc;

import akka.grpc.ProtobufSerializer;
import akka.grpc.javadsl.GoogleProtobufSerializer;

/**
 * The greeting service definition.
 */
public interface GreeterService {
  
  /**
   * Sends a greeting
   */
  java.util.concurrent.CompletionStage<example.myapp.helloworld.grpc.HelloReply> sayHello(example.myapp.helloworld.grpc.HelloRequest in);
  
  /**
   * Comment spanning
   * on several lines
   */
  java.util.concurrent.CompletionStage<example.myapp.helloworld.grpc.HelloReply> itKeepsTalking(akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed> in);
  
  /**
   *
   * C style comments
   */
  akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloReply, akka.NotUsed> itKeepsReplying(example.myapp.helloworld.grpc.HelloRequest in);
  
```

bjaglin commented:
Just discovered that protobuf comments like
```
    /* C style comments
     * on several
     * lines */
```
produce non-valid java/scala source files, investigating

bjaglin commented:
Updated the test and the formatting to be more resilient:

GreeterService.scala
```
/**
 *#services
 * The greeting service definition.
 *
 */
trait GreeterService {
  
  /**
   * Sends a greeting
   *
   */
  def sayHello(in: example.myapp.helloworld.grpc.HelloRequest): scala.concurrent.Future[example.myapp.helloworld.grpc.HelloReply]
  
  /**
   * Comment spanning
   * on several lines
   *
   */
  def itKeepsTalking(in: akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed]): scala.concurrent.Future[example.myapp.helloworld.grpc.HelloReply]
  
  /**
   *
   * C style comments
   *
   */
  def itKeepsReplying(in: example.myapp.helloworld.grpc.HelloRequest): akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloReply, akka.NotUsed]
  
  /**
   * C style comments
   * on several lines
   * with non-empty heading/trailing line 
   */
  def streamHellos(in: akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed]): akka.stream.scaladsl.Source[example.myapp.helloworld.grpc.HelloReply, akka.NotUsed]
```

GreeterService.java
```
/**
 * The greeting service definition.
 *
 */
public interface GreeterService {
  
  /**
   * Sends a greeting
   *
   */
  java.util.concurrent.CompletionStage<example.myapp.helloworld.grpc.HelloReply> sayHello(example.myapp.helloworld.grpc.HelloRequest in);
  
  /**
   * Comment spanning
   * on several lines
   *
   */
  java.util.concurrent.CompletionStage<example.myapp.helloworld.grpc.HelloReply> itKeepsTalking(akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed> in);
  
  /**
   *
   * C style comments
   *
   */
  akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloReply, akka.NotUsed> itKeepsReplying(example.myapp.helloworld.grpc.HelloRequest in);
  
  /**
   * C style comments
   * on several lines
   * with non-empty heading/trailing line 
   */
  akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloReply, akka.NotUsed> streamHellos(akka.stream.javadsl.Source<example.myapp.helloworld.grpc.HelloRequest, akka.NotUsed> in);
```

raboof commented:
Thanks for this great contribution!
